pkg/strutil/merge.go

```diff
@@ -4,42 +4,54 @@
 package strutil
 
 import (
+	"context"
 	"sort"
 	"strings"
 )
 
 // MergeSlices merges a set of sorted string slices into a single ones
 // while removing all duplicates.
-func MergeSlices(a ...[]string) []string {
+func MergeSlices(ctx context.Context, a ...[]string) ([]string, error) {
 	if len(a) == 0 {
-		return nil
+		return nil, nil
 	}
 	if len(a) == 1 {
-		return a[0]
+		return a[0], nil
 	}
 	l := len(a) / 2
-	return mergeTwoStringSlices(MergeSlices(a[:l]...), MergeSlices(a[l:]...))
+	a1, err := MergeSlices(ctx, a[:l]...)
+	if err != nil {
+		return nil, err
+	}
+	a2, err := MergeSlices(ctx, a[l:]...)
+	if err != nil {
+		return nil, err
+	}
+	return mergeTwoStringSlices(ctx, a1, a2)
 }
 
 // MergeUnsortedSlices behaves like StringSlices but input slices are validated
 // for sortedness and are sorted if they are not ordered yet.
-func MergeUnsortedSlices(a ...[]string) []string {
+func MergeUnsortedSlices(ctx context.Context, a ...[]string) ([]string, error) {
 	for _, s := range a {
 		if !sort.StringsAreSorted(s) {
 			sort.Strings(s)
 		}
 	}
-	return MergeSlices(a...)
+	return MergeSlices(ctx, a...)
 }
 
-func mergeTwoStringSlices(a, b []string) []string {
+func mergeTwoStringSlices(ctx context.Context, a, b []string) ([]string, error) {
```

Comment:
I kinda dont understand why this function needs to check for context errors here. Its just processing two string slices.

Reply:
When slices are big enough, this merge can take a lot of time and in cortex we saw it was processing even though request has timed out already, wasting the resources. I think this case is valid for Thanos as well.